### PR TITLE
CSUB-1075: feat(cli): check min bond amount before calling the actual extrinsic

### PR DESCRIPTION
**Pull request by DylanVerstraete**
# Description of proposed changes

See: https://gluwa.atlassian.net/browse/CSUB-1075

Cli will now check for the min validator bond amount before submitting the bond extrinsic, if the amount is not enough it will return an error to the client

---

Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

### CODE REVIEW

**Comment by DylanVerstraete** [comment]
@atodorov not sure why on testnet the minNominatorBond and minValidatorBond are different, can you please explain me why? Should I also take this into account here?

File: cli/src/test/integration-tests/bond.test.ts [hunk 1]
```diff
@@ -133,4 +134,30 @@ describe('bond', () => {
...
+
+            const zero = new BN(0);
+            const oldBalance = await getBalance(caller.address, api);
+            expect(oldBalance.bonded.toString()).toBe(zero.toString());
+            expect(oldBalance.locked.toString()).toBe(zero.toString());
+
+            try {
+                CLI('bond --amount 50');
+            } catch (error: any) {
+                expect(error.exitCode).toEqual(1);
```
Comment (atodorov) [inline cli/src/test/integration-tests/bond.test.ts:155]: missing assert on the actual error message, see other test scenarios for example.

File: cli/src/test/integration-tests/bond.test.ts [hunk 2]
```diff
@@ -133,4 +134,30 @@ describe('bond', () => {
...
+        process.env.PROXY_ENABLED === undefined ||
+            process.env.PROXY_ENABLED === 'no' ||
+            (process.env.PROXY_ENABLED === 'yes' && process.env.PROXY_SECRET_VARIANT === 'valid-proxy'),
+        'should get error if bonding specified amount < min bond amount',
+        async () => {
+            // set staking config min bond amount
+            await setMinBondConfig(api, 100);
+
+            const zero = new BN(0);
+            const oldBalance = await getBalance(caller.address, api);
```
Comment (atodorov) [inline cli/src/test/integration-tests/bond.test.ts:148]: `oldBalance` -> `balance`

File: cli/src/lib/staking/bond.ts [hunk 3]
```diff
@@ -26,6 +26,14 @@ export async function bond(
     if (extra) {
         bondTx = api.tx.staking.bondExtra(amountInMicroUnits.toString());
     } else {
+        // Get min bond amount
+        const min_bond_amount = await api.query.staking.minValidatorBond();
+
+        // Should atleast bond the min validator bond amount on initial bond
+        if (BigInt(amount.toString()) < min_bond_amount.toNumber() * 1e18) {
```
Comment (atodorov) [inline cli/src/lib/staking/bond.ts:33]: `1e18` -> `MICROUNITS_PER_CTC`, no ?

Comment (atodorov) [reply cli/src/lib/staking/bond.ts:33]: On a side note I see that both `lib/constants.ts` and `lib/balance/index.ts` export the MICROUNITS_PER_CTC constant. We need only one of them. I think `lib/constants.ts` is used more widely.

**Review by atodorov — CHANGES_REQUESTED** [review]
Some changes required.

**Comment by atodorov** [comment]
> @atodorov not sure why on testnet the minNominatorBond and minValidatorBond are different

If you mean "different from one another" then the answer is because they are different settings and their values have been decided by stakeholders.

If you mean something else please clarify . 

> Should I also take this into account here?

`minNominatorBond` is related to nominating validators, not becoming a validator. The cli commands don't expose the option for nominating (that is the Staking Dashboard) so this can be ignored IMO.

File: cli/src/lib/staking/bond.ts (hunk 3 above)
Comment (DylanVerstraete) [reply cli/src/lib/staking/bond.ts]: :+1:

File: cli/src/test/integration-tests/bond.test.ts [hunk 4]
```diff
@@ -133,4 +134,31 @@ describe('bond', () => {
...
+
+            try {
+                CLI('bond --amount 50');
+            } catch (error: any) {
+                expect(error.exitCode).toEqual(1);
+                expect(error.stderr).toContain('Amount to bond must be at least the minimum validator bond amount');
+            }
+
+            // revert to 0 again
+            await setMinBondConfig(api, 0);
```
Comment (atodorov) [inline cli/src/test/integration-tests/bond.test.ts:160]: Note: if this test scenario fails for whatever reason before reaching this line the minimum bond config will stay in place and possible break other sceanrios which would be executed afterwards. 

Ideally this call should happen in `afterEach()` to reduce fragility of the test suite.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor change requested.

**Review by atodorov — COMMENTED** [review]
Q: will there be follow-up PRs for `wizard` and `validate` commands ?

**Comment by DylanVerstraete** [comment]
> Q: will there be follow-up PRs for `wizard` and `validate` commands ?

For tests? 

Because the functionality implemented will work for those 2 commands since they call the bond function

**Comment by DylanVerstraete** [comment]
> Q: will there be follow-up PRs for wizard and validate commands ?

You mean `bond` and `wizard` commands? `validate` doesn't do bonding

File: cli/src/lib/staking/bond.ts [hunk 5]
```diff
@@ -26,6 +26,14 @@ export async function bond(
     if (extra) {
         bondTx = api.tx.staking.bondExtra(amountInMicroUnits.toString());
     } else {
+        // Get min bond amount
+        const min_bond_amount = await api.query.staking.minValidatorBond();
+
+        // Should atleast bond the min validator bond amount on initial bond
+        if (BigInt(amount.toString()) < min_bond_amount.toNumber() * MICROUNITS_PER_CTC) {
+            throw new Error('Amount to bond must be at least the minimum validator bond amount');
```
Comment (atodorov) [inline cli/src/lib/staking/bond.ts:34]: Error message can be improved by showing the exact value.

File: cli/src/lib/staking/bond.ts [hunk 6]
```diff
@@ -26,6 +26,14 @@ export async function bond(
     if (extra) {
         bondTx = api.tx.staking.bondExtra(amountInMicroUnits.toString());
     } else {
+        // Get min bond amount
+        const min_bond_amount = await api.query.staking.minValidatorBond();
```
Comment (atodorov) [inline cli/src/lib/staking/bond.ts:30]: nit pick: min_bond_amount -> minBondAmount, or better yet minValidatorBond.

File: cli/src/test/integration-tests/bond.test.ts [hunk 7]
```diff
@@ -133,4 +137,28 @@ describe('bond', () => {
...
+            // set staking config min bond amount
+            await setMinBondConfig(api, 100);
+
+            const zero = new BN(0);
+            const balance = await getBalance(caller.address, api);
+            expect(balance.bonded.toString()).toBe(zero.toString());
+            expect(balance.locked.toString()).toBe(zero.toString());
+
+            try {
+                CLI('bond --amount 50');
```
Comment (atodorov) [inline cli/src/test/integration-tests/bond.test.ts:156]:
```
$ node dist/cli.js bond -u ws://127.0.0.1:9955 --amount 50 -r Stash 
✔ Specify a seed phrase for the caller account … ****************************************************************************
Creating bond transaction...
Reward destination: Stash
Amount: 50.000000000000000000 CTC
✔ Continue? … yes
Amount: 50000000000000000000
/home/gluwa/creditcoin-3.0/creditcoin3/cli/node_modules/bn.js/lib/bn.js:6
    if (!val) throw new Error(msg || 'Assertion failed');
                    ^

Error: Number can only safely store up to 53 bits
    at assert (/home/gluwa/creditcoin-3.0/creditcoin3/cli/node_modules/bn.js/lib/bn.js:6:21)
    at u128.toNumber (/home/gluwa/creditcoin-3.0/creditcoin3/cli/node_modules/bn.js/lib/bn.js:547:7)
    at bond (/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/lib/staking/bond.js:20:57)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Command.bondAction (/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/commands/staking/bond.js:37:26)

Node.js v20.11.1
```


Not sure why the second assertion doesn't fail in CI but that's preventing me from exercising these changes locally.

Comment (atodorov) [reply cli/src/test/integration-tests/bond.test.ts:156]: FTR I have configured:

image

which is the same as Testnet.

**Review by atodorov — CHANGES_REQUESTED** [review]
minor changes requested however I am seeing a traceback locally,.

File: cli/src/lib/staking/bond.ts (hunk 6 above)
Comment (DylanVerstraete) [reply cli/src/lib/staking/bond.ts:30]: I'm not used to writing js anymore :D

File: cli/src/test/integration-tests/bond.test.ts (hunk 7 above)
Comment (DylanVerstraete) [reply cli/src/test/integration-tests/bond.test.ts:156]: Hmm, I will check

**Comment by atodorov** [comment]
> > Q: will there be follow-up PRs for `wizard` and `validate` commands ?
> 
> For tests?
> 

Both for tests and for the `validate` command itself. 

> Because the functionality implemented will work for those 2 commands since they call the bond function



`wizard` command will need tests: 

```
$ node dist/cli.js wizard -u ws://127.0.0.1:9955 -r Stash --amount 1000
🧙 Running staking wizard...
✔ Specify a seed phrase for the caller account … ****************************************************************************
Using the following parameters:
💰 Stash account: 5EEoJZYPpTYGX1t3shaQbqhRmqvdvoaTycY1gJ4TB8spDgPp
🪙 Amount to bond: 1000.000000000000000000 CTC
🎁 Reward destination: Stash
📡 Node URL: ws://127.0.0.1:9955
💸 Commission: 0
🔐 Blocked: No
✔ Continue? … yes
Sending bond transaction...
Amount: 1000000000000000000000
/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/lib/staking/bond.js:23
            throw new Error(`Amount to bond must be at least: ${amountMsg.toString()} CTC (min validator bond amount)`);
                  ^

Error: Amount to bond must be at least: 20000 CTC (min validator bond amount)
    at bond (/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/lib/staking/bond.js:23:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async bondRoutine (/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/commands/staking/wizard.js:164:30)
    at async Command.<anonymous> (/home/gluwa/creditcoin-3.0/creditcoin3/cli/dist/commands/staking/wizard.js:78:13)

Node.js v20.11.1
[gluwa@fedora cli]$ echo $?
1
```


`validate` (staking.validate) checks MinValidatorBond explicitly and we need to check this on the cli side b/c it's easy to end up in a situation where validate will fail when trying to send the extrinsic. For example:

```
$ node dist/cli.js bond -u ws://127.0.0.1:9955 --amount 21000
✔ Specify a seed phrase for the caller account … ****************************************************************************
Creating bond transaction...
Reward destination: Staked
Amount: 21000.000000000000000000 CTC
✔ Continue? … yes
Amount: 21000000000000000000000
Sending transaction...
Transaction included at blockHash 0xfa485f06d0e3718299841a9e78ae881c016fb366e84e1230e3e55cba668f1785


$ node dist/cli.js unbond -u ws://127.0.0.1:9955 --amount 2000
✔ Specify a seed phrase for the caller account … ****************************************************************************
Sending transaction...
Transaction included at blockHash 0x31a8bcd35e9c11181196336f9ae573fcf4f8db89a65fc54a00be19c3a9b0597d


$ node dist/cli.js status -u ws://127.0.0.1:9955 --substrate-address 5EEoJZYPpTYGX1t3shaQbqhRmqvdvoaTycY1gJ4TB8spDgPp
Validator 5EEoJZYPpTYGX1t3shaQbqhRmqvdvoaTycY1gJ4TB8spDgPp:
┌────────────────┬───────────────────────────────────────────────────────┐
│ Status         │                                                       │
├────────────────┼───────────────────────────────────────────────────────┤
│ Bonded         │ Yes                                                   │
├────────────────┼───────────────────────────────────────────────────────┤
│ Validating     │ No                                                    │
├────────────────┼───────────────────────────────────────────────────────┤
│ Waiting        │ No                                                    │
├────────────────┼───────────────────────────────────────────────────────┤
│ Active         │ No                                                    │
├────────────────┼───────────────────────────────────────────────────────┤
│ Can withdraw   │ No                                                    │
├────────────────┼───────────────────────────────────────────────────────┤
│ Next unlocking │ 2000.000000000000000000 CTC in 17 minutes, 25 seconds │
└────────────────┴───────────────────────────────────────────────────────┘

$ node dist/cli.js set-keys -r -u ws://127.0.0.1:9955 
✔ Specify a seed phrase for the caller account … ****************************************************************************
Sending transaction...
Transaction included at blockHash 0xac5613056e974016814302793e727ce30f55d5f095fc3315436308458ccdcb26


$ node dist/cli.js validate -u ws://127.0.0.1:9955 --commission 5 
✔ Specify a seed phrase for the caller account … ****************************************************************************
Creating validate transaction...
Creating validate transaction with params:
Comission: 50000000
Blocked for new nominators: false
Sending transaction...
Transaction failed with error: "staking.InsufficientBond: Cannot have a validator or nominator role, with value less than the minimum defined by governance (see `MinValidatorBond` and `MinNominatorBond`). If unbonding is the intention, `chill` first to remove one's role as validator/nominator." at block 0xa05f1199858ce04a6137a38145c77c9d9c239aabb8e76944154006770e9b1503
```

File: cli/src/test/integration-tests/wizard.test.ts [hunk 8]
```diff
@@ -92,4 +93,27 @@ describe('wizard', () => {
...
+        'should error when not bonding enough',
+        async () => {
+            // set min bond amount to 500
+            const minValidatorBond = 500;
+            // set staking config min bond amount
+            await setMinBondConfig(api, minValidatorBond);
+
+            try {
+                CLI('wizard --amount 900');
+            } catch (error: any) {
```
Comment (atodorov) [inline cli/src/test/integration-tests/wizard.test.ts:110]: this doesn't look right. It is using a bond amount > 500 which isn't supposed to fail, yet the test reports all good. That is not what I expected.

File: cli/src/test/integration-tests/wizard.test.ts [hunk 9]
```diff
@@ -92,4 +93,27 @@ describe('wizard', () => {
...
+    testIf(
+        process.env.PROXY_ENABLED === undefined ||
+            process.env.PROXY_ENABLED === 'no' ||
+            (process.env.PROXY_ENABLED === 'yes' && process.env.PROXY_SECRET_VARIANT === 'valid-proxy'),
+        'should error when not bonding enough',
+        async () => {
+            // set min bond amount to 500
+            const minValidatorBond = 500;
+            // set staking config min bond amount
+            await setMinBondConfig(api, minValidatorBond);
```
Comment (atodorov) [inline cli/src/test/integration-tests/wizard.test.ts:106]: missing cleanup in `afterEach`

File: cli/src/test/integration-tests/validate.test.ts [hunk 10]
```diff
@@ -122,5 +126,38 @@ describe('validate', () => {
...
+            async () => {
+                // set min bond amount to 500
+                const minValidatorBond = 500;
+                // set staking config min bond amount
+                await setMinBondConfig(api, minValidatorBond);
+
+                // unbond 500 (from 900 initially)
+                const unbondResult = CLI('unbond --amount 500');
+                expect(unbondResult.exitCode).toEqual(0);
+                expect(unbondResult.stdout).toContain('Transaction included at block');
```
Comment (atodorov) [inline cli/src/test/integration-tests/validate.test.ts:144]: Lines 140-145 are redundant. 

The amount already bonded is 900 CTC, just make the minBondAmount higher, e.g. 1000 and call validate.

File: cli/src/test/integration-tests/validate.test.ts [hunk 11]
```diff
@@ -122,5 +126,38 @@ describe('validate', () => {
...
+        testIf(
+            process.env.PROXY_ENABLED === undefined ||
+                process.env.PROXY_ENABLED === 'no' ||
+                (process.env.PROXY_ENABLED === 'yes' && process.env.PROXY_SECRET_VARIANT === 'valid-proxy'),
+            'should error if not enough bonded',
+            async () => {
+                // set min bond amount to 500
+                const minValidatorBond = 500;
+                // set staking config min bond amount
+                await setMinBondConfig(api, minValidatorBond);
```
Comment (atodorov) [inline cli/src/test/integration-tests/validate.test.ts:139]: missing cleanup in afterEach

**Review by atodorov — CHANGES_REQUESTED** [review]
Soe changes requested, performing manual tests b/c one of the automated scenarios didn't fail as expected.

File: cli/src/test/integration-tests/wizard.test.ts (hunk 8 above)
Comment (DylanVerstraete) [reply cli/src/test/integration-tests/wizard.test.ts:110]: I've noticed the accounts aren't cleaned up properly, every test is just bonding more to that account

File: cli/src/test/integration-tests/validate.test.ts (hunk 10 above)
Comment (DylanVerstraete) [reply cli/src/test/integration-tests/validate.test.ts:144]: Good idea

File: cli/src/test/integration-tests/wizard.test.ts (hunk 8 above)
Comment (atodorov) [reply cli/src/test/integration-tests/wizard.test.ts:113]: Found the failing execution here (failed with timeout though):
https://github.com/gluwa/creditcoin3/actions/runs/8393365071/job/22988730430

maybe github action statuses did not get updated. Will check statuses again after current jobs finish.

File: cli/src/lib/staking/bond.ts [hunk 12]
```diff
@@ -15,7 +15,9 @@ export async function bond(
 ) {
     console.log(`Amount: ${amount.toString()}`);
 
-    if (BigInt(amount.toString()) < BigInt(MICROUNITS_PER_CTC)) {
+    const precision = BigInt(MICROUNITS_PER_CTC);
+
+    if (BigInt(amount.toString()) < precision) {
```
Comment (atodorov) [inline cli/src/lib/staking/bond.ts:20]: Unrelated change, no? 

Variable `precision` isn't used anywhere else.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor change requested, waiting on tests to complete.

File: cli/src/lib/staking/bond.ts (hunk 12 above)
Comment (DylanVerstraete) [reply cli/src/lib/staking/bond.ts:20]: Made it a const in this file

Comment (atodorov) [reply cli/src/lib/staking/bond.ts:20]: I meant that this change isn't related to the actual scope of 1075, it's something extra which ideally should be proposed separately.

File: cli/src/test/integration-tests/validate.test.ts [hunk 13]
```diff
@@ -73,6 +74,11 @@ describe('validate', () => {
             expect(result.stdout).toContain('Transaction included at block');
         });
 
+        afterEach(async () => {
+            // set default min bond config to 0
+            await setMinBondConfig(api, 0);
+        }, 90_000);
```
Comment (atodorov) [inline cli/src/test/integration-tests/validate.test.ts:80]: This block is valid only in the scope of `describe('when ALREADY bonded', () ` while the actual test that manipulates minValidatorBond is outside of this scope. 

This needs to be moved on line 41.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: cli/src/lib/staking/bond.ts (hunk 12 above)
Comment (DylanVerstraete) [reply cli/src/lib/staking/bond.ts:20]: It's just a type alias to ` BigInt(MICROUNITS_PER_CTC)` which is now used in two places in this file

**Comment by DylanVerstraete** [comment]
Extracted runtime here: https://github.com/gluwa/creditcoin3/pull/288

Will come back to the tests in a later stage

**Comment by atodorov** [comment]
Moving into Draft b/c of the test suite. Will pick it up later.

**Review by atodorov — APPROVED** [review]

**Review by zacharyfrederick — COMMENTED** [review]
LGTM. Did this really drop the smart contract integration test time by an hour?

**Comment by atodorov** [comment]
> Did this really drop the smart contract integration test time by an hour?

No. This CI job has lots of fluctuations but this particupar PR has nothing to do with it.

**Pull request closed without merging**